wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/KubernetesToolToWorkspaceApplier.java

```diff
@@ -86,24 +94,113 @@ public void apply(
             || OPENSHIFT_TOOL_TYPE.equals(k8sTool.getType()),
         format("Plugin must have `%s` or `%s` type", KUBERNETES_TOOL_TYPE, OPENSHIFT_TOOL_TYPE));
 
-    String recipeFileContent = retrieveContent(k8sTool, contentProvider);
+    String toolContent = retrieveContent(k8sTool, contentProvider);
 
-    List<HasMetadata> list = unmarshal(k8sTool, recipeFileContent);
+    List<HasMetadata> toolObjects = new ArrayList<>();
+    fillIn(toolObjects, unmarshalToolObjects(k8sTool, toolContent));
 
     if (!k8sTool.getSelector().isEmpty()) {
-      list = SelectorFilter.filter(list, k8sTool.getSelector());
+      toolObjects = SelectorFilter.filter(toolObjects, k8sTool.getSelector());
     }
 
-    estimateCommandsMachineName(workspaceConfig, k8sTool, list);
+    estimateCommandsMachineName(workspaceConfig, k8sTool, toolObjects);
 
-    applyEntrypoints(k8sTool.getEntrypoints(), list);
+    applyEntrypoints(k8sTool.getEntrypoints(), toolObjects);
 
-    RecipeImpl recipe =
-        new RecipeImpl(k8sTool.getType(), YAML_CONTENT_TYPE, asYaml(k8sTool, list), null);
+    provisionEnvironment(workspaceConfig, k8sTool, toolObjects);
+  }
+
+  /**
+   * Provision default environment with the specified kubernetes/openshift tool.
+   *
+   * <p>If there is already a default environment with kubernetes/openshift then content will be
+   * updated with new list.
+   *
+   * @param workspaceConfig workspace where recipe should be provisioned
+   * @param k8sTool kubernetes/openshift tool that should be provisioned
+   * @param toolObjects parsed objects of the specified tool
+   * @throws DevfileRecipeFormatException if exception occurred during existing environment parsing
+   * @throws DevfileRecipeFormatException if exception occurred during kubernetes object
+   *     serialization
+   * @throws DevfileException if any other exception occurred
+   */
+  private void provisionEnvironment(
+      WorkspaceConfigImpl workspaceConfig, Tool k8sTool, List<HasMetadata> toolObjects)
+      throws DevfileException, DevfileRecipeFormatException {
+    String defaultEnv = workspaceConfig.getDefaultEnv();
+    EnvironmentImpl environment = workspaceConfig.getEnvironments().get(defaultEnv);
+    if (environment == null) {
+      RecipeImpl recipe =
+          new RecipeImpl(k8sTool.getType(), YAML_CONTENT_TYPE, asYaml(k8sTool, toolObjects), null);
+      String envName = k8sTool.getName();
+      workspaceConfig.getEnvironments().put(envName, new EnvironmentImpl(recipe, emptyMap()));
+      workspaceConfig.setDefaultEnv(envName);
+    } else {
+      RecipeImpl envRecipe = environment.getRecipe();
+
+      // check if it is needed to update recipe type since
+      // kubernetes tool is compatible with openshift but not vice versa
+      if (OPENSHIFT_TOOL_TYPE.equals(k8sTool.getType())
+          && KubernetesEnvironment.TYPE.equals(envRecipe.getType())) {
+        envRecipe.setType(OpenShiftEnvironment.TYPE);
```

Comment:
What error occurs if this is not done/is it a user error? If so, I would prefer to notify the user rather than fix it ourselves.

This is kind of a hole in my understanding though, so it may not apply.

Reply:
Well, a user defines Devfile and does not care about phases of Devfile to WorkspaceConfig converting. And if user Defined two tools, one K8s and another OS, then we should provision OpenShift recipe.
Hope my explanation makes things clearer, if not feel free to ask more.

Reply:
:+1: